service/history/replicationTaskFetcher.go

```diff
@@ -0,0 +1,123 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"github.com/uber/cadence/common/service/config"
+	"time"
+
+	"github.com/uber/cadence/.gen/go/cadence/workflowserviceclient"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+)
+
+const (
+	fetchTaskRequestTimeout = 10 * time.Second
+	requestChanBufferSize   = 1000
+)
+
+type (
+	replicationTaskFetcher struct {
+		sourceCluster string
+		config        *config.FetcherConfig
+		logger        log.Logger
+		remotePeer    workflowserviceclient.Interface
+		requestChan   chan *request
+		done          chan struct{}
+	}
+)
+
+func newReplicationTaskFetcher(logger log.Logger, sourceCluster string, config *config.FetcherConfig, sourceFrontend workflowserviceclient.Interface) *replicationTaskFetcher {
+	return &replicationTaskFetcher{
+		config:        config,
+		logger:        logger,
+		remotePeer:    sourceFrontend,
+		sourceCluster: sourceCluster,
+		requestChan:   make(chan *request, requestChanBufferSize),
+		done:          make(chan struct{}),
+	}
+}
+
+func (f *replicationTaskFetcher) Start() {
+	for i := 0; i < f.config.RpcParallelism; i++ {
+		go f.fetchTasks()
+	}
+	f.logger.Info("Replication task fetcher started.", tag.ClusterName(f.sourceCluster), tag.Counter(f.config.RpcParallelism))
+}
+
+func (f *replicationTaskFetcher) Stop() {
+	close(f.done)
+	f.logger.Info("Replication task fetcher stopped.", tag.ClusterName(f.sourceCluster))
+}
+
+func (f *replicationTaskFetcher) fetchTasks() {
+	jitter := backoff.NewJitter()
+	timer := time.NewTimer(jitter.JitDuration(time.Duration(f.config.AggregationIntervalSecs)*time.Second, f.config.TimerJitter))
+	defer timer.Stop()
+
+	requestByShard := make(map[int32]*request)
+	for {
+		select {
+		case request := <-f.requestChan:
+			requestByShard[*request.token.ShardID] = request
```

Comment:
GetShardID()

Reply:
could you also add a comment here
```
this section accumulates requests to fetch replication tasks from remote
```

Reply:
i think here there should be a sanity check, i.e.
```
if req, ok := requestByShard[request.token.GetShardID()]; ok {
  close(req.respChan)
}
requestByShard[request.token.GetShardID()] = request
```

Reply:
done